File: arena-pytest/src/arena_pytest/exec/executable_component.py
from enum import Enum
from typing import Any, Dict, List, Optional, Tuple, Union

from arena_pytest.ffi._ffi_children import children_for_ffi
from arena_pytest.ffi._ffi_readiness import ReadinessCheckEntry, readiness_checks_for_ffi
from arena_pytest.support._identifier import build as _build_identifier
from arena_pytest.readiness import HttpReadinessCheck, TcpReadinessCheck


class BuildTool(Enum):
    CARGO = "cargo"
    MAVEN = "maven"
    GRADLE = "gradle"
    DOTNET = "dotnet"
    MAKE = "make"
    CMAKE = "cmake"

    @staticmethod
    def custom(command: str, args: List[str]) -> Dict[str, Any]:
        return {"command": command, "args": args}


class ExecutableComponentBuilder:
    def __init__(self, name: str = ""):
        self._config: Dict[str, Any] = {
            "type": "exec",
            "identifier": _build_identifier("arena-executable-component", name),
            "env_vars": {},
            "runtime_args": [],
        }
        self._readiness_checks: List[ReadinessCheckEntry] = []
        self._children: List[Any] = []

    def with_executable_path(self, path: str) -> "ExecutableComponentBuilder":
        self._config["executable_path"] = path
        return self

    def with_source_path(self, path: str) -> "ExecutableComponentBuilder":
        self._config["source_path"] = path
        return self

    def with_build_tool(self, build_tool: BuildTool) -> "ExecutableComponentBuilder":
        self._config["build_tool"] = build_tool.value
        return self

    def with_build_tool_custom(self, command: str, args: List[str]) -> "ExecutableComponentBuilder":
        self._config["build_tool"] = BuildTool.custom(command, args)
        return self

    def with_env_var(self, key: str, value: str) -> "ExecutableComponentBuilder":
        self._config["env_vars"][key] = value
        return self

    def with_runtime_arg(self, name: str, value: str) -> "ExecutableComponentBuilder":
        self._config["runtime_args"].append({"name": name, "value": value})
        return self

    def with_readiness_check(
        self,
        check: Union[HttpReadinessCheck, TcpReadinessCheck],
        target: str,
        timeout_ms: int = 10_000,
    ) -> "ExecutableComponentBuilder":
        self._readiness_checks.append((check, target, timeout_ms))
        return self

    def with_child_components(self, children: List[Any]) -> "ExecutableComponentBuilder":
        self._children.extend(children)
        return self

    def build(self) -> "ExecutableComponent":
        return ExecutableComponent(
            dict(self._config),
            readiness_checks=list(self._readiness_checks),
            children=list(self._children),
        )

    def _for_ffi(self) -> Dict[str, Any]:
        d = dict(self._config)
        children = children_for_ffi(self._children)
        if children:
            d["children"] = children
        return d


class ExecutableComponent:
    def __init__(
        self,
        config: Dict[str, Any],
        readiness_checks: Optional[List[ReadinessCheckEntry]] = None,
        children: Optional[List[Any]] = None,
    ):
        self._config = config
        self._readiness_checks = readiness_checks or []
        self._children = children or []

    def _for_ffi(self) -> Dict[str, Any]:
        d = dict(self._config)
        rc = readiness_checks_for_ffi(self._readiness_checks)
        if rc:
            d["readiness_checks"] = rc
        children = children_for_ffi(self._children)
        if children:
            d["children"] = children
        return d

File: arena-ffi/src/component/executable/executable_component.rs
use arena::Component;
use arena_executable_component::builder::ExecutableComponentBuilder;
use arena_executable_component::executable_component::ExecutableComponent;
use arena_executable_component::BuildTool;
use serde::Deserialize;
use std::collections::HashMap;

use crate::healthcheck::{HttpReadinessCheck, ReadinessCheckConfig, TcpReadinessCheck};
use crate::runtime_args::RuntimeArgConfig;

#[derive(Debug, Deserialize)]
#[serde(untagged)]
pub enum BuildToolConfig {
    Simple(String),
    Custom { command: String, args: Vec<String> },
}

#[derive(Debug, Deserialize)]
pub struct ExecutableComponentConfig {
    pub identifier: String,
    pub executable_path: String,
    #[serde(default)]
    pub source_path: Option<String>,
    #[serde(default)]
    pub build_tool: Option<BuildToolConfig>,
    #[serde(default)]
    pub env_vars: Option<HashMap<String, String>>,
    #[serde(default)]
    pub runtime_args: Option<Vec<RuntimeArgConfig>>,
    #[serde(default)]
    pub readiness_checks: Option<Vec<ReadinessCheckConfig>>,
    #[serde(default)]
    pub readiness_check_url: Option<String>,
    #[serde(default)]
    pub cpu_profile_output: Option<String>,
    #[serde(default)]
    pub cpu_profile_auto_open: bool,
    #[serde(default)]
    pub cpu_profile_hotspots: bool,
}

pub fn build(config: &ExecutableComponentConfig) -> Result<Component, String> {
    let mut builder = ExecutableComponent::builder(&config.identifier)
        .with_executable_path(&config.executable_path);

    if let Some(source_path) = &config.source_path {
        builder = builder.with_source_path(source_path);
    }
    let mut cpu_profile_supported = false;
    if let Some(build_tool) = &config.build_tool {
        let bt = build_tool_from_config(build_tool)?;
        cpu_profile_supported = matches!(&bt, BuildTool::Cargo | BuildTool::Maven | BuildTool::Gradle | BuildTool::Python);
        builder = builder.with_build_tool(bt);
    }
    if let Some(env_vars) = &config.env_vars {
        for (k, v) in env_vars {
            builder = builder.with_env_var(k, v);
        }
    }
    if let Some(runtime_args) = &config.runtime_args {
        for arg in runtime_args {
            builder = builder.with_runtime_arg(&arg.name, &arg.value);
        }
    }
    if let Some(output_path) = &config.cpu_profile_output {
        if !cpu_profile_supported {
            return Err(format!(
                "{}: cpu_profile_output requires build_tool to be one of cargo, maven, gradle, or python (got {:?})",
                config.identifier, config.build_tool
            ));
        }
        builder = builder.with_cpu_profile(output_path);
        if config.cpu_profile_auto_open {
            builder = builder.with_cpu_profile_auto_open();
        }
        if config.cpu_profile_hotspots {
            builder = builder.with_hotspots();
        }
    }
    builder = apply_readiness_checks(builder, &readiness_checks_for(config));

    Ok(Box::new(builder.build()))
}

fn readiness_checks_for(config: &ExecutableComponentConfig) -> Vec<ReadinessCheckConfig> {
    let mut v = config.readiness_checks.clone().unwrap_or_default();
    if v.is_empty() {
        if let Some(ref url) = config.readiness_check_url {
            v.push(ReadinessCheckConfig::Http {
                target: url.clone(),
                timeout_ms: 10_000,
            });
        }
    }
    v
}

fn apply_readiness_checks(
    mut builder: ExecutableComponentBuilder,
    checks: &[ReadinessCheckConfig],
) -> ExecutableComponentBuilder {
    for c in checks {
        builder = match c {
            ReadinessCheckConfig::Http { target, timeout_ms } => {
                builder.with_readiness_check_timeout(
                    HttpReadinessCheck::new(),
                    target.as_str(),
                    *timeout_ms,
                )
            }
            ReadinessCheckConfig::Tcp { target, timeout_ms } => {
                builder.with_readiness_check_timeout(
                    TcpReadinessCheck::new(),
                    target.as_str(),
                    *timeout_ms,
                )
            }
        };
    }
    builder
}

fn build_tool_from_config(spec: &BuildToolConfig) -> Result<BuildTool, String> {
    match spec {
        BuildToolConfig::Simple(s) => match s.as_str() {
            "cargo" => Ok(BuildTool::Cargo),
            "maven" => Ok(BuildTool::Maven),
            "gradle" => Ok(BuildTool::Gradle),
            "dotnet" => Ok(BuildTool::Dotnet),
            "make" => Ok(BuildTool::Make),
            "cmake" => Ok(BuildTool::CMake),
            "python" => Ok(BuildTool::Python),
            other => Err(format!("unknown build_tool '{other}'")),
        },
        BuildToolConfig::Custom { command, args } => Ok(BuildTool::Custom {
            command: command.clone(),
            args: args.clone(),
        }),
    }
}
